File: ruby/allergies/allergies.rb
class Allergies
  ITEMS = %w(eggs peanuts shellfish strawberries tomatoes chocolate pollen cats)

  attr_reader :score

  def initialize(score)
    @score = score
  end

  def allergic_to?(item)
    list.include?(item)
  end

  def list
    ITEMS.select.with_index do |item, index|
      score & 2**index > 0
    end
  end
end

File: go/allergies/allergies.go
package allergies

import "math"

var allergens = []string{"eggs", "peanuts", "shellfish", "strawberries", "tomatoes", "chocolate", "pollen", "cats"}

func Allergies(score int) (result []string) {
	for i, allergen := range allergens {
		if int(math.Pow(2, float64(i)))&score >= 1 {
			result = append(result, allergen)
		}
	}

	return
}

func AllergicTo(score int, allergen string) bool {
	for i, item := range allergens {
		if item == allergen && int(math.Pow(2, float64(i)))&score >= 1 {
			return true
		}
	}
	return false
}
